### PR TITLE
Data Template Validation

### PR DESCRIPTION
**Pull request by karacasoft**
## About

Don't know why, but decided to work on this stuff...

The frontend for the default template stuff is broken if it relies on the API side of the project. Because it was relying on putting null to the template field. I think we should not allow null template field. A group should be created with a 'default' template. I'll leave this for now and return to building android.

### Changes

- Data Templates are now checked for compliance to the template docs
- Posts are now checked if they are fitting with their data template

## Visuals
image

### CODE REVIEW

File: api/interestr/api/serializers.py
Comment (smddzcy) [inline api/interestr/api/serializers.py:57]: Nice grammar 🐐

**Review by smddzcy — DISMISSED** [review]
Code looks fine - thanks for the tests 👍  I'll leave the logic review to @ThoAppelsin who is more passionate about the structure of our data templates 😄

**Comment by karacasoft** [comment]
Waiting for @ThoAppelsin 's review... :+1:

**Comment by ThoAppelsin** [comment]
I will do this, and many other work that I've missed, tomorrow. Computer Vision homework had taken a hold of me.

**Comment by karacasoft** [comment]
It's ok. This can stay here for a while, since it has very low priority

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [inline api/interestr/api/serializers.py]: *Checked it: We currently do pass an empty string as a response for when a multiple-select is submitted without any selections.*

Should a response really be always non-empty?

Not that I am really fond of it, but currently we process the post inputs on the front-end and turn them into static strings of responses, hence we have just two fields, `question` and `response`, although the template schema is nowhere that so simple. We have our data structured there.

As a result of this, we, for example, concatenate the checked items on a multiple-select field, separated by commas, and store that as the `response`. How do we convert a multiple-select field when there are no options selected?

We already might be presenting something special such as `(none)`, or maybe not. If we are not, then we maybe can. However, I think it is better to just allow an empty response, and display that as a "(none)" or whatever we wish on our front-ends.

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [inline api/interestr/api/serializers.py]: The validator checks for the uniqueness of the questions within the `post_data`, and then the set-equality of them against the field legends within `template`. The check gives the `post_data` the flexibility to have its question-response pairs in a mixed order.

I would say that the order is also important and should be kept consistent against the data template.  Index-wise correspondences could be made obligatory on this validator. This would put things in order, and make the validation process computationally less intensive.

Additionally: The `%s` placeholders on the format string is left unformatted.

File: api/interestr/api/tests.py
Comment (ThoAppelsin) [inline api/interestr/api/tests.py]: The function name implies that this is a test for/against post creation with no relation to a template, but the post data within does have a related template. I think it should be renamed into something like `test_create_post_disagreeing_with_its_template` or something.

This could very well be a test for a post with no relation to a template.

File: api/interestr/api/tests.py
Comment (ThoAppelsin) [inline api/interestr/api/tests.py]: It could be better if the second field had the same legend (already ok), but with a completely different structure, i.e. different type and inputs. It is not clear whether this test fails for the fields being completely identical, or only because their legends are the same.

**Review by ThoAppelsin — CHANGES_REQUESTED** [review]
My last two comments aren't so critical, however I think the first two needs to get addressed before this gets merged.

Especially the first one either needs to be discussed or changed.

File: api/interestr/api/serializers.py
Comment (karacasoft) [reply api/interestr/api/serializers.py]: > Should a response really be always non-empty?

Good point. I'll remove the check.

File: api/interestr/api/serializers.py
Comment (karacasoft) [reply api/interestr/api/serializers.py]: No index information is given in the data structure. Also, order is not enforced in the documentation. Will not fix this.

> The %s placeholders on the format string is left unformatted.

I'll fix this though.

File: api/interestr/api/tests.py
Comment (karacasoft) [reply api/interestr/api/tests.py]: Ok :+1:

File: api/interestr/api/tests.py
Comment (karacasoft) [reply api/interestr/api/tests.py]: The test is sufficient. If we have to test if our system allows multiple fields with completely identical fields that only have different legends, we should write a different test case for that.

**Comment by karacasoft** [comment]
I tl;dr'd your paragraphs. Please, let me know if I missed a point

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [inline api/interestr/api/serializers.py]: List of valid types does not have the newly introduced `"select"` type of fields. Since this pull request comes after the #174, it should include this.

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [inline api/interestr/api/serializers.py]: No errors are caught in the case of value (fields) being empty. A template may not have its fields completely empty. It currently does on the web front-end, but it shouldn't really.

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [reply api/interestr/api/serializers.py]: Documentation has absolutely nothing about the posts. I had been told that the posts do not require one for being simple and intuitive, or something like that.

Documentation has something about the templates. It mentions about the arrays on the template fields. Arrays are inherently ordered.

File: api/interestr/api/tests.py
Comment (ThoAppelsin) [reply api/interestr/api/tests.py]: The title `test_create_multiple_field_same_legend_template` misleads the reviewer that this test is for testing specifically for the identical field legends. As it is, it could very well be testing (and failing) for having two fields with identical types.

**Review by ThoAppelsin — CHANGES_REQUESTED** [review]
I have updated the [data templates' fields documentation](../wiki/Documentation-for-the-JSON-Structure-of-Data-Template-Fields-and-Posts) to reflect the changes with the PR #174.

The validators currently invalidate functionalities that are introduced with that update. I have made some further reviews on them, and written some replies to the replies.

The validators are, of course, not complete, but I am not looking for a complete/thorough validation either. I just want the parts that could cause problems to be fixed.

File: api/interestr/api/tests.py
Comment (karacasoft) [reply api/interestr/api/tests.py]: > It could very well be testing (and failing) for having two fields with identical types.

Why?

File: api/interestr/api/serializers.py
Comment (karacasoft) [reply api/interestr/api/serializers.py]: Ok

File: api/interestr/api/serializers.py
Comment (karacasoft) [reply api/interestr/api/serializers.py]: True. Good point.

File: api/interestr/api/tests.py
Comment (ThoAppelsin) [reply api/interestr/api/tests.py]: Let me put it this way: How do we know that we get a status code of 400 because the field legends are the same, and not for another reason?

Comment (karacasoft) [reply api/interestr/api/tests.py]: We do not know. And probably will never if we do not add another case I mentioned. You cannot test everything at once.

Comment (ThoAppelsin) [reply api/interestr/api/tests.py]: This test is too broad. It can be made more specific to honor its name, to a degree that I cannot anymore complain about.

I mean, either change the name of the test into `test_create_multiple_identical_field_template`, or one of the two fields' type and inputs into something else. The name of the test and the test itself are not in accordance as it is now.

**Comment by ThoAppelsin** [comment]
It is very rude and uncooperative of you to not read the reviews I've made thoroughly. You do seem to have read all my points, though, so it is rather only just very rude of you.

**Comment by karacasoft** [comment]
@ThoAppelsin Let's do an interactive lesson about writing less stuff:

> Checked it: We currently do pass an empty string as a response for when a multiple-select is submitted without any selections.
>
> Should a response really be always non-empty?
>
>~~Not that I am really fond of it, but currently we process the post inputs on the front-end and turn them into static strings of responses, hence we have just two fields, question and response, although the template schema is nowhere that so simple. We have our data structured there.~~
>
> ~~As a result of this,~~ we ~~, for example,~~ concatenate the checked items on a multiple-select field, separated by commas, and store that as the response. ~~How do we convert a multiple-select field when there are no options selected?~~
>
> We ~~already might be presenting~~ _may present_ something special such as (none)~~, or maybe not. If we are not, then we maybe can.~~ However, I think it is better to just allow an empty response, and display that as a "(none)" or whatever we wish on our front-ends.

Here I have read one of your long responses and converted into a much shorter passage that conveys absolutely the same meaning.

Let's do a part by part examination:

> Checked it: We currently do pass an empty string as a response for when a multiple-select is submitted without any selections.

No problem so far. You are stating something I might not know. Can be worked on to be shortened, but it's ok.

> Should a response really be always non-empty?

Good question.

> Not that I am really fond of it, but currently we process the post inputs on the front-end and turn them into static strings of responses, hence we have just two fields, question and response, although the template schema is nowhere that so simple. We have our data structured there.

You are just writing what we already know. I mean if I'm working on validating this stuff, I think I should know how our data structure is. 3 lines long text (on desktop) that serves no purpose.

> As a result of this, we, for example, concatenate the checked items on a multiple-select field, separated by commas, and store that as the response. How do we convert a multiple-select field when there are no options selected?

Good stuff, but mostly not related to this issue. We were talking about leaving response field empty. No need to waste other people's time with this. But we need to have connection with the next sentence. I would probably combine these two paragraphs into one, but for simplicity I'll just simplify them both.

> We already might be presenting something special such as (none), or maybe not. If we are not, then we maybe can. However, I think it is better to just allow an empty response, and display that as a "(none)" or whatever we wish on our front-ends.

The only relevant part here is about allowing empty responses being better. So we just remove whatever that's not about it. But adding a special "(none)" text idea is cool and can be shared with others to be implemented later. So we leave that here, too.

The point is, I should be able to see my mistakes directly and do something about it as soon as I see your response. You don't have to prove something in your reviews. And I don't want to read an essay every single time I see your name in my mailbox.

AND, NO don't write a ~300 paragraphs answer for this. I will not read it. I'll simply return to my job now.

**Comment by ThoAppelsin** [comment]
You can never know how many lines it takes for the other side to fully comprehend what's going on in your mind in one-shot. I tend to take the safe route.

That long comment of mine which you took as an example got the following response from you:

> Good point. I'll remove the check.

Easy-peasy. Took me only about 10-20 minutes to write that and get things done. Much better than starting off with 3 lines and then having a 3-hours-long discussion about something that we all should have learned about 3 years ago as we took the Cmpe 160. See: https://github.com/bounswe/bounswe2017group5/pull/186#discussion_r157476474

**Comment by ThoAppelsin** [comment]
This is just nonsense.

I will do the changes I recommended myself and seek for a review.

**Review by ThoAppelsin — DISMISSED** [review]

**Comment by ThoAppelsin** [comment]
I have just made a ~47 line commit considering the reviews I had made and re-opened this. This thing is valuable, and only needed a couple of fixes. It would be better if someone else reviewed it again, before we merge it.

**Comment by karacasoft** [comment]
Added one more fix in `validate_fields`. Please check

File: api/interestr/api/serializers.py
Comment (ThoAppelsin) [inline api/interestr/api/serializers.py]: Great, thanks!

**Review by smddzcy — APPROVED** [review]
Approved, again...

**Pull request merged**
